### PR TITLE
[IGA] 5p Hierarchic Shell

### PR DESCRIPTION
**Pull request by tteschemacher**
Hey all,

this PR adds the developments of @MichaelLoibl on the 5 parameter shell, which was developed in Stuttgart by Oesterle under Bischoff. It includes a set of tests implemented by @rickyaristio .

**Description**
5p hierarchic shell based upon the Kirchhoff-Love shell.

### CODE REVIEW

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp
Comment (rath3t) [inline applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp:1259]: This would also benefit from #7385 to use a more appropriate naming.

File: applications/IgaApplication/tests/linear_beam_thick_p4_nCP5/linear_beam_thick_p4_nCP5_test_results.json
Comment (rath3t) [inline applications/IgaApplication/tests/linear_beam_thick_p4_nCP5/linear_beam_thick_p4_nCP5_test_results.json:9]: this would also reduce pollution here

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp
Comment (tteschemacher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp:1259]: I agree!

File: applications/IgaApplication/tests/linear_beam_thick_p4_nCP5/linear_beam_thick_p4_nCP5_test_results.json
Comment (rath3t) [inline applications/IgaApplication/tests/linear_beam_thick_p4_nCP5/linear_beam_thick_p4_nCP5_test_results.json:17]: and would reduce pollution here

File: applications/IgaApplication/tests/linear_beam_thick_p4_nCP5/linear_beam_thick_p4_nCP5_test_results.json
Comment (rath3t) [inline applications/IgaApplication/tests/linear_beam_thick_p4_nCP5/linear_beam_thick_p4_nCP5_test_results.json]: this wolud also reduce pollution here

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp
Comment (rath3t) [inline applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp:355]: Tih would also benefit from #7385. Since the more resonable variable names can be simply defined

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp
Comment (philbucher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp:1259]: How would it change?

File: applications/IgaApplication/tests/nonlinear_beam_thick_p2_nCP22/shell_5p_project_parameters.json
Comment (philbucher) [inline applications/IgaApplication/tests/nonlinear_beam_thick_p2_nCP22/shell_5p_project_parameters.json]:
```suggestion
```

this was completely removed

File: applications/IgaApplication/tests/test_modelers.py
Comment (philbucher) [inline applications/IgaApplication/tests/test_modelers.py]:
```suggestion
```

not needed any more

File: applications/IgaApplication/tests/linear_beam_thick_p4_nCP5/linear_beam_thick_p4_nCP5_test_results.json
Comment (philbucher) [reply applications/IgaApplication/tests/linear_beam_thick_p4_nCP5/linear_beam_thick_p4_nCP5_test_results.json]: @tteschemacher why do you write this out?
=> you can specify the components X and Y then Z won't appear in the file

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h]:
```suggestion
        const ProcessInfo& rCurrentProcessInfo
```

Comment (philbucher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h]: @tteschemacher I strongly suggest to use the new interfaces, this will be deprecated & removed soon => #7381

its not much work, basically adding some const

Comment (philbucher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h]: also I would move the implementation to the cpp

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h:599]: missing implementation

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp
Comment (philbucher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp:1259]: I thought some more, consider also that by default we use `ROTATION` for rotational dofs

if you use some other then you will have to make some more adjustments to the solver which might not be trivial

=> even if it doesn't fit 100% your intention I would carefully consider using some different variable, this can have nasty consequences!

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp]:
```suggestion
```

no longer needed, now keys are assigned at compiletime and ensured to be non-zero

File: applications/IgaApplication/tests/linear_scordelis_p4_nCP10/shell_5p_project_parameters.json
Comment (philbucher) [inline applications/IgaApplication/tests/linear_scordelis_p4_nCP10/shell_5p_project_parameters.json:110]: @tteschemacher can you please use a submodelpart here to reduce the file size of the result files?

Usually it is not necessary to check every single node, if sth goes wrong you will also know if you check a subset :)

**Review by philbucher — COMMENTED** [review]
Some comments 

I am not blocking but please reduce the file sizes as I pointed out

In order to avoid having big files in the git history (which is already 10x the size of the code!) please consider [squashing](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/merging-a-pull-request#merging-a-pull-request-on-github) this commit instead of merging

File: applications/IgaApplication/tests/test_modelers.py
Comment (tteschemacher) [reply applications/IgaApplication/tests/test_modelers.py]: Thanks. This file comes from #7087 . I changed it there. As soon as #7087 is merged, I will merge the master, here. Then it is disappeared.

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp
Comment (tteschemacher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp]: cheers!

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h
Comment (tteschemacher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h]: yep. process infos are const now. const can safe lives.

Comment (tteschemacher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h]: > also I would move the implementation to the cpp

I agree and disagree. To be honest this could be something which is implemented already in the base element class. It is repeated everywhere and just unnecessary lines of code...

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp
Comment (tteschemacher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp:1259]: We can quite easily use the variable with using the auxiliary_variables_list. However, having the additional third dimension is a bit a mess.

File: applications/IgaApplication/tests/linear_scordelis_p4_nCP10/shell_5p_project_parameters.json
Comment (tteschemacher) [reply applications/IgaApplication/tests/linear_scordelis_p4_nCP10/shell_5p_project_parameters.json:110]: the file size is not even that heavy. However, it seems that the whole simulation is already too big for the CI.

Do all application have the same run time of tests? On my machine it takes only very short in comparison to the structural mechanics application.

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp
Comment (philbucher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp:1259]: That will only add the dofs
There are still more things like Scheme, ConvCrits etc which are optimized to work with Disp (and Rot)

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h
Comment (philbucher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h:599]: still missing

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h
Comment (philbucher) [inline applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h:593]: missing implemetation

**Review by philbucher — COMMENTED** [review]
ok for me but please reduce the sizes of the result files by checking only a few nodes.
we are having more and more problems with this

File: applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h
Comment (tteschemacher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h:599]: Hehe, it's actually not that straight away, to be honest, I'd prefer to keep this for future work. I'll soon add up the internal stresses and stuff and there'll be a reconsideration anyways. Thusly, i prefer to postpone this. In iga we're never using that anyways...

Comment (philbucher) [reply applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.h:599]: > Thusly, i prefer to postpone this

ok for me but I think this is still important

Not having this means you cannot restart which makes IGA basically useless for Clusters where you have to restart due to runtime limitations
Also some real life cases of IGA with embedded FSI will have to be done on the cluster so I would add it soon!

**Comment by tteschemacher** [comment]
> 
> 
> ok for me but please reduce the sizes of the result files by checking only a few nodes.
> we are having more and more problems with this

Is this that important? I mean before reducing the file size I'd probably just remove them completely, because right now it is not that trivial to just check certain points.

**Comment by philbucher** [comment]
> > ok for me but please reduce the sizes of the result files by checking only a few nodes.
> > we are having more and more problems with this
> 
> Is this that important? I mean before reducing the file size I'd probably just remove them completely, because right now it is not that trivial to just check certain points.

yes Kratos grows out of bounds completely :)

you are using the standard json check process with what seems to be the main ModelPart ("IgaModelPart")
Why cannot you use a submodelpart? This would be the way to go usually

**Comment by tteschemacher** [comment]
because the only submodelpart are supports, which are indeedly no good measurement points.

**Comment by philbucher** [comment]
then why don't you create a separate SubModelPart with a few interesting/important points where you check the results?
This is what is usually done in structural and other apps

**Comment by tteschemacher** [comment]
This is not that easy with IGA. I am not purely defining points on the surface or can define model parts within the mdpa. I would always require to ask for data with a certain context. #7545 would help in this specific context, but it doesn't look like it got big supports.

Would you prefer me to remove the tests?

**Comment by philbucher** [comment]
approving after discussing in person
in a nutshell it is currently not possible to easily subdivide the geomerty further, might be in the future

**Review by philbucher — APPROVED** [review]

**Comment by tteschemacher** [comment]
swweeeet

**Pull request merged**
